clippy_lints/src/direct_method_call.rs

```diff
@@ -0,0 +1,89 @@
+#![allow(clippy::redundant_clone)]
+
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::fn_def_id;
+use clippy_utils::source::snippet_with_context;
+use rustc_errors::Applicability;
+use rustc_hir::Expr;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Suggests a better pattern of code, when a trait is used explicitly.
+    /// ### Why is this bad?
+    /// It's poorly readable.
+    /// ### Example
+    /// ```rust
+    /// f32::floor(4.5f32);
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// (4.5f32).floor();
+    /// ```
+    #[clippy::version = "1.66.0"]
+    pub DIRECT_METHOD_CALL,
+    complexity,
+    "Needlessly using explicit traits"
+}
+
+declare_lint_pass!(DirectMethodCall => [DIRECT_METHOD_CALL]);
+
+// 'X::Y(Z) -> Z.Y()' When Z implements X
+
+impl LateLintPass<'_> for DirectMethodCall {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let Some(fnid) = fn_def_id(cx, expr) {
+            match fnid.as_local() {
+                None => {
+                    // TODO: IMPLEMENT FUNCTION THAT AREN'T LOCAL
```

Comment:
The HIR only exists for the current crate, so there's no way to go from a foreign `DefId` to `HirId` or a `rustc_hir::FnDecl`

There is `TyCtxt::fn_sig` that gives similar info for foreign and local functions, but it doesn't contain whether the self type is implicit. For that you could try [`TyCtxt::opt_associated_item`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.opt_associated_item)